src/components/application_manager/src/policies/policy_handler.cc

```diff
@@ -334,39 +333,45 @@ bool PolicyHandler::LoadPolicyLibrary() {
       return std::shared_ptr<PolicyManager>();
     }
 
-    void* const dl_handle = dlopen(kLibrary.c_str(), RTLD_LAZY);
+    void* const dl_policy_handle = dlopen(kLibrary.c_str(), RTLD_LAZY);
 
-    if (!dl_handle) {
+    if (!dl_policy_handle) {
       SDL_LOG_ERROR("An error occurs while calling dlopen");
       return std::shared_ptr<PolicyManager>();
     }
 
     typedef PolicyManager* (*CreateManager)(logger::Logger*);
-
-    CreateManager create_manager =
-        reinterpret_cast<CreateManager>(dlsym(dl_handle, "CreateManager"));
-
-    PolicyManager* const policy_manager_raw =
-        create_manager ? create_manager(&logger::Logger::instance()) : nullptr;
-
-    if (!policy_manager_raw) {
-      if (dlclose(dl_handle)) {
-        SDL_LOG_ERROR("An error occurs while calling dlclose");
-      }
-      SDL_LOG_ERROR(kLibraryNotLoadedMessage);
+    typedef void (*DeleteManager)(PolicyManager*);
+
+    CreateManager create_manager = reinterpret_cast<CreateManager>(
+        dlsym(dl_policy_handle, "CreateManager"));
+    char* error_string = dlerror();
+    if (NULL != error_string) {
+      SDL_LOG_WARN(error_string);
+      dlclose(dl_policy_handle);
       return std::shared_ptr<PolicyManager>();
     }
 
-    std::shared_ptr<PolicyManager> policy_manager{
-        policy_manager_raw, [dl_handle](PolicyManager* pm) noexcept {
-          typedef void (*DeleteManager)(PolicyManager*);
-          DeleteManager delete_manager = reinterpret_cast<DeleteManager>(
-              dlsym(dl_handle, "DeleteManager"));
-          if (delete_manager) {
-            delete_manager(pm);
-          }
-          dlclose(dl_handle);
-        }};
+    DeleteManager delete_manager = reinterpret_cast<DeleteManager>(
+        dlsym(dl_policy_handle, "DeleteManager"));
+
+    auto policy_destroyer = [delete_manager,
+                             dl_policy_handle](PolicyManager* policy_manager) {
+      SDL_LOG_DEBUG("Delete Policy Manager");
+      delete_manager(policy_manager);
+      dlclose(dl_policy_handle);
+    };
+
+    std::shared_ptr<PolicyManager> policy_manager;
+    error_string = dlerror();
+    if (NULL == error_string) {
+      policy_manager = std::shared_ptr<PolicyManager>(
+          create_manager(&logger::Logger::instance()), policy_destroyer);
+    } else {
+      SDL_LOG_WARN(error_string);
+      dlclose(dl_policy_handle);
+      return std::shared_ptr<PolicyManager>();
+    }
 
     policy_manager->set_listener(this);
```

Comment:
My suggestion:

    DeleteManager delete_manager = reinterpret_cast<DeleteManager>(
        dlsym(dl_policy_handle, "DeleteManager"));
    error_string = dlerror();
    if (NULL != error_string) {
      SDL_LOG_WARN(error_string);
      dlclose(dl_policy_handle);
      return std::shared_ptr<PolicyManager>();
    }

    auto policy_destroyer = [delete_manager,
                             dl_policy_handle](PolicyManager* policy_manager) {
      SDL_LOG_DEBUG("Delete Policy Manager");
      delete_manager(policy_manager);
      dlclose(dl_policy_handle);
    };

    std::shared_ptr<PolicyManager> policy_manager(
        create_manager(&logger::Logger::instance()), policy_destroyer);

    policy_manager->set_listener(this);